### ISSUE TITLE
How to identify books with missing items?

### ISSUE THREAD

**Issue by bilcorry** [openaudible/openaudible#1088]
All of my books are converted to M4B, except 2 of them.  How do I know which 2 books are missing the M4B file?  There's a search for M4B, but not a negative search for missing M4B.

Same with PDFs.  All of my books with PDFs have been downloaded except 1 book.  How can I identify the 1 book that is missing its PDF?

Thanks.

**Comment by bilcorry** [comment]
For the missing converted files, the "state" is green for all books.  Sorting by the first column has no effect since all books have a green status.  Is there a way to have the app update the status indicator by double-checking the files on disk?

For the missing PDF, selecting all, then choosing "Get PDF" didn't work because "Get PDF" is grayed out, I can't choose that option.  I was able to find the one book missing a PDF by manually looking.  While I select that one book and try to do "Get PDF", it still doesn't work because "Get PDF" is grayed out.   

I noticed for the book missing the PDF, it does have a "file name" listed next to the PDF label:

     PDF:   HTTP/1.1 404 Not Found

Maybe it thinks it has a PDF because it has a "filename"?  The books.json file has a URL for the PDF, but the PDF is missing on the Audible website, which is why it can't be downloaded.  Maybe it's correct then that the "Get PDF" feature is grayed out since it can't be downloaded, but if it works that way, it does prevent me from finding the books missing the PDF.

I'm still not sure which two books are missing the M4B files, if there's another way to identify them other than manually looking, that would be appreciated as I have thousands of books.

**Comment by openaudible** [comment]
Items can also be green if they have an MP3 (any co. If you prefer all M4B, you can try clicking on "MP3 files" to see which books have mp3s and delete the mp3s. Then re-encode as m4b. But if you have mp3s from other sources (not audible) we don't recommend deleting those.. and we don't support going from mp3->m4b (since you lose quality).

Usually the PDF is a bad link from audible. Go to the audible page for the book and see if there is a pdf. Not sure where the filename for the pdf came frame.. While we don't recommend it, you could quit openaudible, make a backup of your books.json, and carefully edit the pdf_link for that book.

Will figure out a way to add a full path for each of the converted files to the CSV export so it will be easier to tell what files you have. But that will be in a future beta.

**Comment by bilcorry** [comment]
The PDF is missing at Audible, which is why it's missing in OpenAudible.  But the only reason I know that is because I was able to find the book manually in OpenAudible, then go to Audible's site and see that the PDF is missing at Audible.  So everything is working in OpenAudible except there isn't a way to search for books with missing PDFs.  My feature request is that you add a capability to search for books with missing PDFs.

As for the 2 books with the missing M4B files, I just realized I'm also missing the AAX file too.  What I can't understand is that if I remove a M4B file, OpenAudible detects its missing and auto-converts the AAX file.  If I remove a AAX file, OpenAudible detects it's missing and auto-downloads the AAX file from Audible.  So my thinking is the two books must be missing the AAX file at Audible, but since I don't know which two and all books have a green status, I can't investigate it further.  My feature request is that you add a capability to search for books with missing AAX and M4B files.

And as an aside, none of my books have MP3.  I only have AAX and M4B.

**Comment by openaudible** [comment]
Found a similar problem with one of my PDFs. 

I think you're saying you have say, 500 books, and 498 M4B, 0 MP3 files in the Library Status panel. So hard for you to tell where the missing files are. 

Make sure you're logged in. 

Click on the "Ignored Books" to make sure you didn't ignore some books. 

I will fix the Book Info Panel to not show a pdf filename if the file doesn't exist.  (Will replace with "Needs download", which should get replaced by an error if downloading it fails.)

I'll improve the export function so you can figure out exactly what is going on..by letting you see all the file paths and meta data.

Hoping to get a beta out soon.. but making some bigger changes.

**Comment by bilcorry** [comment]
Thanks, that's right.  I have 2 more books in my library than AAX/M4B files on disk (which matches what is reported by OpenAudible), but all of my books have a green status.  And I did check Ignored books and I don't have any.

Looking forward to the update.  Thanks for making this product!

**Comment by openaudible** [comment]
New beta released to address your issue and other issues.

You can now negate any search by prepending a ! before the search text.

So you can click "M4B" to show all the books that have m4b. In the search, add a ! before the search and it should show all audiobooks without an m4b. Exporting to CSV will also show more details about the files.

Hope that does it. Let me know

**Comment by bilcorry** [comment]
Thank you for the new beta features.

Installing the beta came with the issue that libraries.json reverted to the normal Windows config of C:\, but my OpenAudible library is on another drive.  I tried setting it in the Preferences, but saving and restarting didn't fix it.  Once I manually edited libraries.json with the correct paths, it no longer tried to download my entire library.  This might be a side effect from my Windows setting in that I have my other drive "protected" by Windows and I have to first manually allow an application access to the drive.  But even after allowing the beta access to the drive, the settings wouldn't take and I had to manually update libraries.json.

For the missing PDFs, "!PDFs" finds the books without PDFs, not the books that should have PDFs but don't.  The new export did allow me to find the book with the missing PDF by using Excel to compare the column with the PDF URL and the PDF path on disk (i.e. if there's a URL but no path on disk, then the PDF is missing).  I still think it'd be easier if there was a specific search for missing PDFs (and AAX, M4B, etc), but this is good enough for me.

For the missing M4Bs, "!M4B Files" returns an empty list.  I wonder if two books have the same title and are sharing an AAX and M4B?  I do have two versions of 20,000 Leagues Under the Sea and saw that there's a "1" appended to one of them, so it seems you have some logic to deal with this issue.

Oh, nevermind.  I used the new book export and found that indeed, there are two books that are using the same AAX and M4B files, which is why they're missing, yet not missing.  For one book, it's the same author and title, but two different narrators, but they are sharing the same AAX and M4B.  I wonder if I delete the AAX and the M4B if both will be downloaded.  I can try if you want?

For the other book, I have three different books all with the same title.  I have two of them sharing the same AAX and M4B, and one has a "1" appended to it.

I'm not sure why/how they ended up sharing AAX other than there were a couple of times where the internet went out and I had to interrupt the download and restart, maybe OpenAudible got confused?

As I proposed above, I could try deleting the AAX and M4B for those books with collisions and see if OpenAudible will correctly download and name the AAX.  Or if you want to try to detect this condition and automatically correct it, I can leave them as-is for now.

Again, thank you for the new features.  I appreciate it!

**Comment by openaudible** [comment]
Sorry for the late reply. 

Yes, sometimes duplicates can get into a library, especially if the location changes or books are added via import. 

If you're sure you still have them in your library, you can back the files up, delete the book(s), do a full rescan, and get them again. Or delete just one book and use the "rename" feature if needed to get rid of a 1 after the book name.

**Comment by bilcorry** [comment]
I'm on the latest beta 10.

I removed the AAX and M4B files for the collisions, and had OpenAudible download them again.  Unfortunately, OpenAudible shared the AAX and M4B files again (meaning two books in my library point to the same AAX and M4B).  I fixed it by editing books.json directly, but wanted you to know there is still a bug in OpenAudible.

One of the books affected is The Small House at Allington, which I have two versions, one is slightly longer than the other.  The AAX downloaded was the longer of the two, so when the shorter of the two went to convert it, it showed it converted more of the book than it's total length (i.e. 23:34:07 of 22:50:00).  That's because it was converting the AAX of the longer version.  Here's a screenshot:

<img width="224" alt="Converting" src="https://user-images.githubusercontent.com/53133/213116916-83240c5f-0e9a-4038-be80-d068c94b965f.png">

It also gave an error at the end of converting because the file already existed (which it did because the longer version of the book was being converted at the same time and finished slightly sooner).

For the other book that had a collision, as it was redownloading the AAX, the status icon wasn't updated and instead still showed as completed.  Here's the screenshot of that:

<img width="403" alt="Status" src="https://user-images.githubusercontent.com/53133/213118308-97eef67e-8163-4ca3-a391-fe97e7040f8b.png">

Anyhow, my issue is resolved, my library and my files all now match.  Thank you for the updated export feature, it made it possible to identify the books that had collisions.

If you want to try to replicate the collision, these are the two books that OpenAudible seems to get caught up on (at least for me):

This one is free:
https://www.audible.com/pd/The-Small-House-At-Allington-Audiobook/B002V8MR0K

This one is free if you have a Plus membership:
https://www.audible.com/pd/The-Small-House-at-Allington-Audiobook/B002V19YMG

Thanks again!
